### PR TITLE
Issue a log.info message with inconsistent SIP representation.

### PR DESCRIPTION
**Pull request by nden**
This is a follow up on #4803. While I prefer issuing a `Warning`, warnings are issued once per session and it is easy to miss the problem in subsequent calls to `wcs.WCS`. To avoid this I implemented it as a `log.info` message. However, I can convert it to a warning message if that's desired.

So please, suggest better wording or let me know if you think `log.info` is too intrusive and I'll change it to a warning. 

_Update on what this PR implements:_

- If SIP coefficients are in the header but `CTYPE` **is missing** " -SIP" a log.info message is printed when the header is read in. The WCS includes SIP distortion.
- If SIP coefficients are in the WCS but `CTYPE` **is missing** "-SIP" a log.info message is printed when `wcs.to_header(relax=True)` is called and "-SIP" is appended to CTYPE.
- If SIP coefficients are in the WCS and `-SIP` **is in** `CTYPE`, "-SIP" is stripped from `CTYPE` when `wcs.to_header()` is called with `relax=None` or `relax=False`.

### CODE REVIEW

**Comment by nden** [comment]
@astrofrog @eteq @antonkoekemoer @mcara

**Comment by pllim** [comment]
also c/c @hcferguson

**Comment by pllim** [comment]
@nden , perhaps a change log under "API change" for 1.2?

File: astropy/wcs/wcs.py [hunk 1]
```diff
@@ -1031,6 +1032,16 @@ def _read_sip_kw(self, header):
 
             del header[str('A_ORDER')]
             del header[str('B_ORDER')]
+            ctype=[header['CTYPE{0}'.format(nax)] for nax in range(1, self.naxis + 1)]
+            if any([ctyp[-4 :] != '-SIP' for ctyp in ctype]):
```
Comment (mcara) [inline astropy/wcs/wcs.py]: It seems that SIP Paper (2008) specifies that '-SIP' is appended to CTYPE1 and CTYPE2. It is not clear if '-SIP' suffix is appended to CTYPE3 if it is a spectral axis.

File: astropy/wcs/wcs.py [hunk 2]
```diff
@@ -1031,6 +1032,16 @@ def _read_sip_kw(self, header):
...
+                message = """
+                         Inconsistent SIP distortion information in header.
+                         Detected SIP coefficients but CTYPE is missing "-SIP" suffix.
+                         Please consider correcting the header.
+                         SIP distortion component was included in the WCS. Use ``wcs_pix2world``
+                         instead of ``all_pix2world`` to ignore all distortion components.
```
Comment (mcara) [inline astropy/wcs/wcs.py]: `wcs_pix2world` turns off not only SIP but also all other (D2IM, CPDIS) distortions which may or may not be what user wants: see https://github.com/astropy/astropy/issues/4803#issuecomment-214573200. Maybe it would be better to tell users to turn off SIP (only) by setting `WCS.sip = None`

**Comment by mcara** [comment]
1. Now log files will be never ending... May I suggest issuing log.debug if this is the only approach to solve issues #4803 and #4761.
2. This warning, as currently implemented, is designed to work only during header readout. It would do nothing to warn users who change CTYPE in memory but do not turn remove ``sip`` from WCS (i.e., if object is modified in memory).
3. It will not warn users if when they are saving the in-memory modified WCS objects to the disk which is what could have helped CANDELS team that their files had non-standard SIP in image headers.

File: astropy/wcs/wcs.py (hunk 1 above)
Comment (nden) [inline astropy/wcs/wcs.py]: @mcara SIP applies to images only.

**Comment by mcara** [comment]
As I have expressed in https://github.com/astropy/astropy/issues/4803#issuecomment-214570886, I now strongly support making `astropy.wcs` support for SIP compliant with the [SIP paper (2008)](http://fits.gsfc.nasa.gov/registry/sip/SIP_distortion_v1_0.pdf) and do use `-SIP` suffix in ``CTYPE`` as a switch for applying SIP corrections. Also, the code should be modified in such a way that if a WCS object was edited in memory such that ``-SIP`` suffix was removed from ``CTYPE`` then ``sip`` attribute is set to ``None``.

**Comment by antonkoekemoer** [comment]
@mcara wrote:

 > 1.   Now log files will be never ending... May I suggest issuing log.debug if this is the only approach to solve issues #4803 and #4761.

The majority of logfiles should not be affected since this info message applies in a rather particular circumstance (ie images with CTYPE missing "-SIP" but with SIP keywords present). I'd agree with @ndev that in this case the user needs to be informed about this issue with log.info, not just in debug mode.

> 2. This warning, as currently implemented, is designed to work only during header readout..

This is the case that triggered the original thread, so it remains relevant, and it would remain useful for this info message to be generated on reading in the header so that the user can be made aware of the issue and take appropriate action.

**Comment by nden** [comment]
@mcara That's right. This is simply giving information, it's not changing behaviour or files.
I think this is what 1) in #4803 was about (doing the minimum non-invasive change) and most who commented agreed this is not a bad thing.

**Comment by eteq** [comment]
I also agree that this should *not* be `log.debug` - `info` is just fine.  The whole point is to inform users about the assumption that Astropy is making here, and `debug` is essentially invisible to users.

One way to lower the log burden, though, would be to add a bool keyword to the initializer like `force_sip`, which would simply hide the warning in the condition it catches, and probably *also* raise an error if the sip coefficents are missing and it's set (It would be counter-intuitive for ``force_sip=True`` to silently parse something that had no sip keywords).

But I agree with @mcara that it might be good to duplicate the check when *writing*.  Although I'm not sure if that's possible given where the writing happens.  If it seems straightforward, I'd say @nden should add it, but if not it's not nearly as critical as on-read-in.

**Comment by nden** [comment]
@eteq @astrofrog I have added a log.info message when writing the WCS to a header. In the process I discovered that when `to_header(relax=False)` or just `to_header()` is called and the SIP keywords are correctly not written out it will still write out `CTYPE` with "-SIP" if present in the original header. This is clearly a bug so in these cases "-SIP" is stripped from CTYPE. I've updated the description above. 

Feel free to merge after reviewing. As I mentioned above the `to_header` method has many corner cases. I don't know who else is familiar with this part of the code so I'm randomly assigning @astrofrog :-). Feel free to assign it to someone else.

File: astropy/wcs/tests/test_wcs.py [hunk 3]
```diff
@@ -914,3 +914,26 @@ def test_no_truncate_using_compare():
...
+    assert(all([ctyp[-4 :] != '-SIP' for ctyp in self.wcs.ctype]))
+    newhdr = w.to_header(key="C")
+    # Test writing header with a different key
+    assert(all([ctyp[-4 :] != '-SIP' for ctyp in self.wcs.ctype]))
+    newhdr = w.to_header(key=" ")
+    # Test writing a primary WCS to header
+    assert(all([ctyp[-4 :] != '-SIP' for ctyp in self.wcs.ctype]))
+    # Test that "-SIP" is kept into CTYPE if relax=True
+    newhdr = w.to_header(relax=True)
+    assert(all([ctyp[-4 :] == '-SIP' for ctyp in self.wcs.ctype]))
```
Comment (astrofrog) [inline astropy/wcs/tests/test_wcs.py]: @nden - in the above tests, can you also add a test of whether one of the SIP coefficients is present or not in the resulting header?

**Comment by astrofrog** [comment]
@nden - as far as you know, are there any scenarios under which we will ever return a header that has SIP coefficients but no ``-SIP`` suffix? I think we should make sure we always produce unambiguous files, hence my suggestion above to check that the presence/absence of SIP keywords correlates with the ``-SIP`` suffix.

**Comment by nden** [comment]
@astrofrog This PR does not change current functionality (except for not writing "-SIP" in `CTYPE` when `relax=False`). I left it as it is because of some opposition to that change.

However, now that I looked at the code in detail, I think it's safe to write out "-SIP" in `CTYPE` in the presence of coefficients without changing anyone's headers. The reason is that the default `to_header()` call does not write out coefficients. One has to specifically ask to have them in the output by setting `relax=True` which means we should add "-SIP" to `CTYPE`.

@antonkoekemoer

File: astropy/wcs/tests/test_wcs.py (hunk 3 above)
Comment (nden) [inline astropy/wcs/tests/test_wcs.py]: Do you mean check for coefficients when `relax=True`?

**Comment by nden** [comment]
If we add "-SIP" to `CTYPE` when `relax=True`, we should change the log.info message as well. Currently it says:

```
Inconsistent SIP distortion information:

                SIP coefficients were detected, but CTYPE is missing "-SIP" suffix,
                therefore the header will be incorrect.

                If distortion should be included in the header please add "-SIP" to CTYPE.

                If distortion should not be included please remove the SIP coefficients.
```

It should say something like:

```
SIP coefficients were detected, but CTYPE is missing "-SIP" suffix.
Adding "-SIP" to CTYPE.
If distortion should not be included please use `to_header(relax=False)`.
```

I've no objection either way.

**Comment by antonkoekemoer** [comment]
@nden @astrofrog
Printing a log.info message when writing the header is good, to say that the header may be inconsistent and the user should either add "-SIP" to CTYPE (if the image is distorted) or else remove the extraneous SIP coefficients (if the image is truly undistorted); I'm also fine with the current wording.

However, changing the actual behaviour, to now add "-SIP" to CTYPE if SIP coefficients are found, makes the assumption that this is what the user intends (ie, assumes the image is in fact still distorted). If the image is now truly undistorted, then these extraneous SIP coefficients should instead be removed and CTYPE should remain without "-SIP" (otherwise, adding "-SIP" to CTYPE for an image that's already distortion-corrected will yield incorrect astrometry although the header will now appear self-consistent).

So, an alternative change in functionality (also making an assumption, but the opposite one) would remove SIP coefficients if CTYPE is missing "-SIP", and would keep CTYPE without "-SIP". This would also produce a self-consistent header, and the message in this case could say something like:

> SIP coefficients were detected, but CTYPE is missing "-SIP" suffix,
> indicating that this image may already have had its distortion removed.
> For consistency, these extraneous SIP coefficients are being removed from header.
> If instead these SIP coefficients should be retained, then "-SIP" must be
> added to CTYPE, and in that case please use `to_header(___________)`.

So, both of these changes in functionality (either adding "-SIP" to CTYPE, or keeping CTYPE without "-SIP" and instead removing SIP coefficients from the header) are making an assumption that this is what the user wants, but in two opposite directions.

While I'm definitely in favour of adding the log.info message to warn the user that the header is inconsistent, and tell the user how to fix it either way, it's less clear that we can anticipate what the user actually intended. Both of these opposite choices can be justified for changing the behaviour, but it would be making an assumption, either way.

So I would be fine with not taking the step of actually changing the default behaviour to either modify CTYPE or remove SIP coefficients. But certainly it's necessary to make it clear to the user that they need to either add "-SIP" or remove the SIP coefficients, and I think the proposed log.info message states this quite well.

**Comment by nden** [comment]
@antonkoekemoer and I talked about this issue and agreed on how to proceed. I've updated the description with the latest changes. To summarize:

- If SIP coefficients are in the input header and "-SIP" is not in CTYPE, a log.info message is printed and the SIP distortion is included in the WCS when the header is read in. (no change of functionality)

- If SIP distortion is in the WCS but "-SIP" is missing in CTYPE, `wcs.to_header(relax=True)` will now append "-SIP" to `CTYPE` in addition to writing out SIP coefficients, writing out a consistent header.
("-SIP" in CTYPE is new.)

- If SIP distortion is in the WCS and "-SIP" is in CTYPE, `wcs.to_header(relax=False)` and `wcs.to_header()` will now strip "-SIP" from `CTYPE` in addition to not writing out SIP coefficients, writing out a consistent header. ("-SIP" not in CTYPE is new.)

I think this is ready to merge (after travis finishes) and I'd like to get it in 1.2.

**Comment by pllim** [comment]
@nden, any chance you can squash the commits and let Travis CI run to make sure tests are okay?

**Comment by nden** [comment]
Hm, what triggered travis cancellation?

**Comment by pllim** [comment]
I cancelled it because `master` was failing, so yours will fail regardless. But `master` is fixed now. Could you please rebase again and push? Thanks!

**Comment by MSeifert04** [comment]
Must have been a [fluke](https://github.com/astropy/astropy/pull/4868#issuecomment-218568086) *cough*

**Comment by pllim** [comment]
Maybe squash the commits while you are at it? Thanks!

**Comment by nden** [comment]
I see. I was wondering if I did something wrong.

**Comment by pllim** [comment]
It's not you, it's me...

**Comment by astrofrog** [comment]
@nden - just to follow up on my (now hidden) comment, I was thinking it would be good to check in ``test_inconsistent_sip`` that when you call to_header, when ``-SIP`` is absent the SIP keywords are also absent (or just one of them for simplicity), and that when ``-SIP`` is included in ``CTYPE``, the SIP keywords are there. Essentially testing that we never end up with an inconsistent header where there is no ``-SIP`` but SIP keywords, or vice versa. Maybe you could just add the test with one coefficient?

**Comment by eteq** [comment]
@nden, both of the remaining commits have a ``[ci skip]`` (the first one due to the squashing), so the tests didn't run.  So if you want to add the test @astrofrog suggested you can do that, otherwise the easiest thing is probably to just remove the ``[skip ci]`` from the second commit and re-push that up.

(Incidentally, this is exactly the sort of reason why I prefer not to squash commits any more than strictly necessary @pllim - but to each their own ;)

**Comment by nden** [comment]
@ astrofrog done:  added a test

**Comment by pllim** [comment]
@eteq , yes, it comes down to preference. But perhaps standard procedures could be established on when to and not to request people to squash commits in Astropy, then at least maintainers can be consistent about it. :sweat_smile:

**Comment by pllim** [comment]
@nden -- one PEP8 failure

```
astropy/wcs/tests/test_wcs.py:969:1: W391 blank line at end of file
```

**Comment by astrofrog** [comment]
@pllim - in general, I ask people to squash if their history includes mess-ups that they removed in subsequent commits, or if it includes merge commits. If the history is clean with a few commits, I usually am fine with that.

**Comment by astrofrog** [comment]
Thanks @nden!

**Comment by eteq** [comment]
👍  - thanks @nden!

**Comment by pllim** [comment]
@astrofrog , "a few" is subjective. For example, 5 commits for a 3-line change is not a few by ratio but it is a few because it is less than 100.

**Comment by keflavich** [comment]
This PR has broken some things I think it should not have.  For example, the header below now crashes with the following error:

```
  File "/Users/adam/repos/spectral-cube/spectral_cube/wcs_utils.py", line 348, in strip_wcs_from_header
    wcsh = hwcs.to_header()
  File "/Users/adam/repos/astropy/astropy/wcs/wcs.py", line 2541, in to_header
    header = self._fix_ctype(header, add_sip=False)
  File "/Users/adam/repos/astropy/astropy/wcs/wcs.py", line 2622, in _fix_ctype
    val = header[kw].strip("-SIP")
  File "/Users/adam/repos/astropy/astropy/io/fits/header.py", line 154, in __getitem__
    card = self._cards[self._cardindex(key)]
  File "/Users/adam/repos/astropy/astropy/io/fits/header.py", line 1726, in _cardindex
    raise KeyError("Keyword %r not found." % keyword)
KeyError: "Keyword 'CTYPE4' not found."
```

In the `_fix_wcs` step, the header keywords should only be replaced if they exist.

```
SIMPLE  =                    T         /
BITPIX  =                   16         /
NAXIS   =                    4         /
NAXIS1  =                  512         /
NAXIS2  =                    1         /
NAXIS3  =                    1         /
NAXIS4  =                    1         /
BLOCKED =                    T         /
BLANK   =                32767         / Blanking value
BSCALE  =  0.4882752364210E-05         /
BZERO   =  0.1805613376200E-01         /
DATAMIN = -0.1419418901205E+00         /
DATAMAX =  0.1780443936586E+00         /
BUNIT   = 'K       '                   /
CTYPE1  = 'FREQ    '                   /
CRVAL1  =  0.0000000000000E+00         / Offset frequency
CDELT1  = -0.1000000119209E+07         / Frequency resolution
CRPIX1  =  0.2565000000000E+03         /
CTYPE2  = 'RA---GLS'                   /
EQUINOX =  0.2000000000000E+04         /
CRVAL2  =  0.2721558333333E+03         /
CDELT2  =  0.0000000000000E+00         /
CRPIX2  =  0.0000000000000E+00         /
CTYPE3  = 'DEC--GLS'                   /
CRVAL3  = -0.2031791666667E+02         /
CDELT3  =  0.0000000000000E+00         /
CRPIX3  =  0.0000000000000E+00         /
CROTA3  =  0.8058610690795-297         /
CTYPE4  = 'STOKES  '                   /
CRVAL4  =      1.0000000000000         /
CDELT4  =      0.0000000000000         /
CRPIX4  =      0.0000000000000         /
TELESCOP= 'SMT-F1M-VL  '               /
OBJECT  = '10074-190   '               /
LINE    = 'HCOP(3-2)   '               / Line name
RESTFREQ=  0.2675576250000E+12         / Rest frequency
VELO-LSR=  0.0000000000000E+00         / Velocity of reference channel
DELTAV  =  0.1120478272438E+04         / Velocity spacing of channels
IMAGFREQ=  0.2795576250000E+12         / Image frequency
TSYS    =  0.2203572998047E+03         / System temperature
OBSTIME =  0.2400000000000E+03         / Integration time
SCAN-NUM=  0.1610000000000E+03         / Scan number
TAU-ATM =  0.1480000019073E+00         / Atmospheric opacity
GAINIMAG=  0.0000000000000E+00         / Image sideband gain ratio
BEAMEFF =  0.1000000000000E+01         / Beam efficiency
FORWEFF =  0.1000000000000E+01         / Image sideband gain ratio
ORIGIN  = 'CLASS-Grenoble  '           /
DATE    = '2011-03-12T00:00:00.000'    / Date written
TIMESYS = 'UTC             '           /
DATE-OBS= '2011-02-28T07:29:39.858'    / Date observed
DATE-RED= '2011-03-02T00:00:00.000'    / Date reduced
ELEVATIO=  0.2904819348342E+02         /  Telescope elevation
AZIMUTH =  0.1453942095413E+03         /  Telescope azimuth
UT      = ' 07:29:39.858'              /  Universal time at start
LST     = ' 19:07:31.684'              /  Sideral time at start
HISTORY SCAN-LIST        20-      20
END
```

**Comment by nden** [comment]
@keflavich Are you sure this is the header that fails? This one has `CTYPE`. And `to_header()` works when `CTYPE` is missing as the example below works:

```
w = wcs.WCS()
w._to_header()
```

**Comment by keflavich** [comment]
@nden sorry, no, my mistake, I grabbed the wrong header.  It's this one:

```
SIMPLE  =                    T         /
BITPIX  =                  -32         /
NAXIS   =                    4         /
NAXIS1  =                   21         /
NAXIS2  =                   30         /
NAXIS3  =                  501         /
NAXIS4  =                    1         /
DATAMIN = -0.1457159280777E+01         /
DATAMAX =  0.6162142753601E+01         /
BLOCKED =                    T         /
BUNIT   = 'K           '               /
CTYPE1  = 'RA---ARC    '               /
CRVAL1  =  0.2466102288723E+03         /
CDELT1  = -0.2083333386675E-02         /
CRPIX1  =  0.1493333352010E+02         /
CROTA1  =  0.0000000000000E+00         /
CTYPE2  = 'DEC--ARC    '               /
CRVAL2  = -0.2440855924977E+02         /
CDELT2  =  0.2083333386675E-02         /
CRPIX2  =  0.9153333173918E+01         /
CROTA2  =  0.0000000000000E+00         /
CTYPE3  = 'VELOCITY    '               /
CRVAL3  =  0.2500000000000E+04         /
CDELT3  = -0.6284131109715E+02         /
CRPIX3  =  0.4030960083008E+03         /
CROTA3  =  0.0000000000000E+00         /
CROTA4  =  0.0000000000000E+00         /
OBJECT  = 'VLA1623A    '               /
RA      =  0.2466102288723E+03         / Right Ascension
DEC     = -0.2440855924977E+02         / Declination
EPOCH   =  0.2000000000000E+04         /
EQUINOX =  0.2000000000000E+04         /
RESTFREQ=  9.3176265000000E+10 / Line rest frequency (Hz)
IMAGFREQ=  9.6172401511998E+10 / Image frequency (Hz)
VLSR    =  2.5000000000000E+03 / LSR velocity (m s-1)
ORIGIN  = 'GILDAS-(LAOG-IRAM) '
DATE    = '04/07/07'                   / Date written
DATE-OBS= '2007-07-04T00:00:00.00'     / Date written
CUNIT3  = 'm/s     '
HISTORY File modified by user 'belloche' with fv  on 2007-07-04T16:32:18
END
```

This may be because of an operation `spectral_cube` is doing to strip out header keywords...

**Comment by nden** [comment]
I see. There are 4 axes , `wcs.naxis = 4` but `CTYPE4` is not in the header (though it is in the wcs as an empty string):
```
 w.wcs.ctype
Out[61]: ['RA---ARC', 'DEC--ARC', 'VELOCITY', '']
```
I'll have a fix shortly. Thanks!
